### PR TITLE
Change requirements airflow max version ~2.4

### PR DESCRIPTION
**Pull request by was-av**
Tested and readme updated

### CODE REVIEW

**Review by bryzgaloff — APPROVED** [review]

**Comment by bryzgaloff** [comment]
@was-av could you please provide a fix for a failed test? Unit test `test_get_hook_defined` fails: `AttributeError: module 'airflow.sensors' has no attribute 'sql'`.

**Comment by was-av** [comment]
@bryzgaloff, yes. The fix is already done and waiting a approve.

File: tests/unit/test_sql_sensor.py [hunk 1]
```diff
@@ -110,8 +110,8 @@ def tearDownClass(cls) -> None:
 
 
 class ClickHouseLegacySqlSensorTestCase(unittest.TestCase):
-    @mock.patch('airflow.sensors.sql.SqlSensor.poke')
-    @mock.patch('airflow.sensors.sql.SqlSensor._get_hook', create=True)
+    @mock.patch('airflow.providers.common.sql.sensors.sql.SqlSensor.poke')
+    @mock.patch('airflow.providers.common.sql.sensors.sql.SqlSensor._get_hook', create=True)
```
Comment (bryzgaloff) [inline tests/unit/test_sql_sensor.py:114]: This is a destructive change which fails for earlier versions of Airflow. When was `sensors.sql` moved to `providers.common.sql`? At which Airflow version?

**Comment by bryzgaloff** [comment]
In general, we should create a dependency matrix with the following relationship:
* Python version.
* Airflow max version.
* ClickHouse plugin max version.

And test them accordingly. In this case users of earlier Python/Airflow versions will be able to pick up a right version of the plugin.

@was-av will you be able to do this as a part of a separate PR? Our workflow config currently is too generic matching improper combinations of the versions.

File: tests/unit/test_sql_sensor.py (hunk 1 above)
Comment (was-av) [reply tests/unit/test_sql_sensor.py:114]: [Airflow 2.4.0 (2022-09-19)](https://airflow.apache.org/docs/apache-airflow/stable/release_notes.html?highlight=airflow+sensors+sql#id178) 

DBApiHook and SQLSensor have moved [(#24836)](https://github.com/apache/airflow/pull/24836)
DBApiHook and SQLSensor have been moved to the apache-airflow-providers-common-sql provider.

**Comment by was-av** [comment]
> Okey. Could you provide more details on how to implement such a matrix?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -34,7 +34,7 @@ jobs:
     strategy:
       matrix:
         python-version: ["3.7", "3.8", "3.9", "3.10"]
-        airflow-version: ["2.1.4", "2.2.5", "2.3.0"]
+        airflow-version: ["2.1.4", "2.2.5", "2.3.0", "2.4.0"]
```
Comment (r3b-fish) [inline .github/workflows/tests.yml:37]: 2.4.1 has been released, you can change to it

File: airflow_clickhouse_plugin/hooks/clickhouse_hook.py [hunk 3]
```diff
@@ -46,7 +48,7 @@ def get_first(self, sql: str, parameters: Optional[dict] = None) -> Optional[Tup
                 return None
 
     def get_pandas_df(self, sql: str):
-        import pandas as pd
+        import pandas as pd # pylint: disable=import-outside-toplevel
```
Comment (bryzgaloff) [inline airflow_clickhouse_plugin/hooks/clickhouse_hook.py:51]: Please do not include your personal pylint noqas into the repo. Everyone uses their own code quality tool.

File: airflow_clickhouse_plugin/hooks/clickhouse_hook.py [hunk 4]
```diff
@@ -62,7 +64,7 @@ def run(
             sql = (sql,)
         with disconnecting(self.get_conn()) as conn:
             last_result = None
-            for s in sql:
+            for s in sql: # pylint: disable=invalid-name
```
Comment (bryzgaloff) [inline airflow_clickhouse_plugin/hooks/clickhouse_hook.py:67]: Drop here too.

File: airflow_clickhouse_plugin/hooks/clickhouse_hook.py [hunk 5]
```diff
@@ -103,7 +105,7 @@ def _log_params(
 _InnerT = TypeVar('_InnerT')
 
 
-class disconnecting(ContextManager, Generic[_InnerT]):
+class disconnecting(ContextManager, Generic[_InnerT]): # pylint: disable=invalid-name
```
Comment (bryzgaloff) [inline airflow_clickhouse_plugin/hooks/clickhouse_hook.py:108]: Drop pylint here too.

File: airflow_clickhouse_plugin/operators/clickhouse_operator.py [hunk 6]
```diff
@@ -1,7 +1,6 @@
-from typing import *
+from typing import Optional, Iterable, Dict, Any, Union
```
Comment (bryzgaloff) [inline airflow_clickhouse_plugin/operators/clickhouse_operator.py:1]: Please do not change anything except from direct purpose of a PR. If you want to perform a refactoring, please make it part of another PR.

File: tests/integration/test_operator.py [hunk 7]
```diff
@@ -1,9 +1,10 @@
 import unittest
 
 from airflow_clickhouse_plugin.operators.clickhouse_operator import ClickHouseOperator
+from tests.integration.util import ClickHouseConnectionEnvVarTestCase
 
 
-class BasicTestCase(unittest.TestCase):
+class BasicTestCase(ClickHouseConnectionEnvVarTestCase):
```
Comment (bryzgaloff) [inline tests/integration/test_operator.py:7]: Have not this test case been working without this patch? I have not seen any problems in the tests output.

**Review by bryzgaloff — CHANGES_REQUESTED** [review]
PR includes too many changes unrelated to an initial purpose. Could you please revert them for the PR to be focused on a version increment only? You may suggest another PR with you code style changes, though I plan to introduce a linter soon.

**Comment by bryzgaloff** [comment]
> Could you provide more details on how to implement such a matrix?

Simply a markdown table. This requires a manual check of different Python/Airflow/plugin versions compatibility. Something like:

Plugin version | Max Airflow version | Max Python version
=== | === | ===
0.8.2 | 2.3.* | 3.10.*
… | … | …

And the .github/workflows/tests.yml versions matrix should list the same combinations, so that we know the tests correspond to a list of versions.

BTW, are any Python or Airflow versions not supported already completely? We do not need to support Python versions earlier than Airflow supports.

**Comment by SergeyBychkov** [comment]
Hello!
Can I offer any help?

**Comment by SergeyBychkov** [comment]
Okey

**Comment by bryzgaloff** [comment]
Guys, good news: Airflow 2.4 support has been added as a part of #50 

@was-av I am going to close this PR. But I will mention you as a contributor on README anyway, thank you for your participation!

**Comment by bryzgaloff** [comment]
Airflow 2.4 support is released: https://pypi.org/project/airflow-clickhouse-plugin/0.9.0/

Though, Airflow 2.1 support is disabled, details: #52
